app/app.go

```diff
@@ -72,12 +100,71 @@ func (app app) runWorkflow(workflow w.Workflow, payload map[string]string) error
 	}
 
 	go func() {
-		log.Printf("Starting workflow with identifier %s, payload: %+v", workflow.Identifier, payload)
+		ctx := log.Fields{
+			"identifier": workflow.Identifier,
+			"payload":    payload,
+		}
+		log.WithFields(ctx).Info("starting workflow")
 		err := runner.Run(workflow, payload)
 		if err != nil {
-			log.Printf("Failed to run workflow: %s", err)
+			log.WithError(err).WithFields(ctx).Error("failed to run workflow")
 		}
 	}()
 
 	return nil
 }
+
+func (app app) runCommand(comm *command.Command) {
```

Comment:
I think this could be a method on a `CommandHandler` "class", instead of it being directly in `app`. The CommandHandler would then "own" the channel, and provide a method to post things to the channel.

This would hide the fact that a channel is used from users of the CommandHandler, so in the future you could use something else, like a message queue.

You could then inject the CommandHandler in the `registry`, instead of posting directly to the channel.

Reply:
I think I follow you here & it sounds better to abstract the channel under a CommandHandler of sorts. But it would still need to call app.runWorkflow in the end, so it can't be out of app unless it has app itself as a dependency (which would be wrong).

Reply:
I was able to abstract usage of channels away under `CommandsHandler` but it looks like app would still be the one to loop over commands as they are sighted and call runCommand on that, so that function would have to stay here. Let me know if that's what you meant or I am missing something else.

[Committed](https://github.com/Automattic/neurobot/commit/31a495445bf152710ec5a4dd2ede89f409473d74) in a separate branch.